### DISCUSSION TITLE
[Question] how to open Virtual Keyboard in window os?

### DISCUSSION THREAD

**Discussion by jjjunehkim** [flet-dev/flet#3583] [Q&A] [answered]
### Question

When trying to use a kiosk without connecting a keyboard,
if I touch(click) on ft.TextField, I want the virtual keyboard to be open in window 11 os, 
 
please explain how to

### ------------------------------------------------------

- [X] I have searched for answers to my question both in the issues and in previous discussions.

**Comment by burhansvural** [comment]
import os
os.system("C:/Windows/System32/osk.exe")

**Reply by jjjunehkim** [reply]
Thank you for answer.
But when I build and run,  virtual keyboard is worked. but cmd window is also opened reapeatly per 5~6 second. 
It is problem.

**Reply by burhansvural** [reply]
> Thank you for answer. But when I build and run, virtual keyboard is worked. but cmd window is also opened reapeatly per 5~6 second. It is problem.

@jjjunehkim 

Write your full codes here so I can study and check them.

**Reply by jjjunehkim** [reply]
I used a command "flet build windows" for make exe file with this code, 
And run exe file. then cmd window is opened reapeatly per 5~6 second.

code is blow. 

import flet as ft
import os

def main(page: ft.Page):

    def change_text(e):
        text.value = textfield.value
        page.update()

    def show_touch_keyboard(e):
        os.system("osk.exe")
        

    textfield = ft.TextField(
                    border_color = ft.colors.BLACK,
                    text_align = "center",
                    on_focus = show_touch_keyboard,
                    on_change = change_text
    )
    
    text = ft.Text(
            value = "",
    )
    

    ct = ft.Column(
            controls = [
                textfield,
                text
            ]
        )


    page.add(ct)

ft.app(target=main)

**Reply by burhansvural** [reply]
@jjjunehkim 
Because the focus event of the textfield is triggered frequently and the virtual keyboard opens every time it is triggered. You can place a button in the application and click it to open the virtual keyboard once. In the meantime, you can search the processes and open the virtual keyboard once if the osk.exe application is not running.

**Reply by jjjunehkim** [reply]
Thank you so much for your solution.

I modified the code a little following your explanation as below, 
Then I created exe file and ran exe file.

Although th cmd window did not open repeatedly, but opened once.

When executing a Python script, the cmd window did not open. 
But when running the created exe file, the cmd window was opened.

**What I really wanted was to put a virtual keyboard into the content of ft.Container, but that was difficult, so I tried to use the Windows virtual keyboard, but a problem occurred with the cmd window opening.**

import flet as ft
import os

def main(page: ft.Page):

    def change_text(e):
        text.value = textfield.value
        page.update()

    def show_touch_keyboard(e):
        textfield.focus()
        os.system("osk.exe")
        
    textfield = ft.TextField(
                    border_color = ft.colors.BLACK,
                    text_align = "center",
                    # on_focus = show_touch_keyboard,
                    on_change = change_text
    )
    
    text = ft.Text(
            value = "",
    )

    button = ft.ElevatedButton(
                text = "click",
                on_click = show_touch_keyboard
    )
    
    ct = ft.Column(
            controls = [
                textfield,
                text,
                button
            ]
        )

    page.add(ct)

ft.app(target=main)

**Reply by burhansvural** [reply, marked as answer]
Then you can write your own virtual keyboard, turn it into a class and use it.
Examples:
https://github.com/topics/virtual-keyboard-python

**Reply by jjjunehkim** [reply]
thank you for your reply.
I will check it .
I don't think it will be easy because my language, Korean, is written differently from English.

**Reply by burhansvural** [reply]
https://www.instructables.com/Using-Python-to-Learn-Non-English-Keyboard-Layouts/

**Reply by jjjunehkim** [reply]
Oh, Thank you so so ~~~~much!!!!
The link is for Korean keyboard!

**Reply by burhansvural** [reply]
yes
